src/compas_fab/robots/reachability_map/reachability_map.py

```diff
@@ -0,0 +1,157 @@
+from compas.data import Data
+from compas.geometry import Frame
+from compas.robots import Configuration
+from compas.geometry import argmax
+
+
+class ReachabilityMap(Data):
+    """The ReachabilityMap describes the reachability of a robot.
+
+    The ReachabilityMap describes the reachability of a robot at certain frames,
+    with valid IK solutions at these frames. The map only makes sense to be
+    calculated with analytic inverse kinematic solvers, as they include all
+    possible solutions to be found.
+
+    Attributes
+    ----------
+    frames : list of :class:`compas.geometry.Frame` or list of list of :class:`compas.geometry.Frame`
+        The frames at which the IK solutions are calculated, either a 1D list or
+        a 2D list.
+    configurations : list of list of :class:`compas.robots.Configuration` or list of list of list of :class:`compas.robots.Configuration`
+        The configurations at the frames, either a 2D list or a 3D list,
+        depending on the dimension of the frames respectively.
+    score : list of int
+        The number of solutions per frame (1D) or per frame list (2D)
+    points : list of :class:`compas.geometry.Point`
+        The points per frame (1D) or per frame list (2D)
+    shape : tuple of int
+        The shape of the frame array
+
+
+    Links
+    -----
+    http://wiki.ros.org/reuleaux
+    """
+
+    def __init__(self, frames=None, configurations=None, name=None):
+        super(ReachabilityMap, self).__init__(name)
+        self.frames = frames or []  # 1D or 2D
+        self.configurations = configurations or []  # 2D or 3D
+
+    def calculate(self, frame_generator, robot, ik_options=None):
+        """Calculates the reachability map.
+
+        Parameters
+        ----------
+        frame_generator : generator
+            A 1D or 2D frame generator to yield :class:`compas.geometry.Frame`.
+            The solutions are saved depending on the dimensions.
+        robot : :class:`compas_fab.robots.Robot`
+            The robot instance for which inverse kinematics is being calculated.
+            This makes only sense if the robot has an analytic inverse kinematic
+            solvers set.
+        ik_options : dict, optional
+            Optional arguments to be passed on to the robot's inverse kinematics
+            function.
+        """
+
+        from compas_fab.backends.exceptions import InverseKinematicsError
```

Comment:
import should go to the top... does it generate errors of circular imports if you do?

Reply:
rhino (ipy) complains

Reply:
i tried again and now it does not complain anymore...

Reply:
sorry, now I know: actually ```tests\api\test_api_completeness.py``` complains
Extract:
```
[autosummary] generating autosummary for: C:\Users\rustr\workspace\compas_fab\docs\api\compas_fab.robots.rst
[autosummary] writing to C:\Users\rustr\workspace\compas_fab\temp\stubs
Failed to import compas_fab.robots.
Possible hints:
* AttributeError: module 'compas_fab' has no attribute 'robots'
* ImportError: cannot import name 'Duration' from partially initialized module 'compas_fab.robots' (most likely due to a circular import) (c:\users\rustr\workspace\compas_fab\src\compas_fab\robots\__init__.py)
```